client/mysql.cc

```diff
@@ -5315,6 +5365,512 @@ static int com_prompt(String *buffer __attribute__((unused)),
+  char *name, *value, *pos;
+  bool is_valid;
+  size_t name_len;
+  uchar *record;
+
+  /* Parse the alias name. */
+  pos= parse_alias_name(line, &name, &is_valid);
+
+  if (!is_valid)
+  {
+    tee_fprintf(stdout, "alias: '%s': invalid alias name\n", name);
+    *error= true;
+  }
+  else
+  {
+    /* TODO: A non-alphanumeric alias name is invalid. */
+
+    /*
+      We have a valid alias name. Lets check if there is a value being
+      assigned to it. (Note: there mustn't be spaces around '='.)
+    */
+    name_len= strlen(name);
+
+    if (*pos == '=')
+    {
+      pos ++;
+      pos= parse_alias_value(pos, &value, &is_valid);
+
+      if (!is_valid)
+      {
+        tee_fprintf(stdout, "alias: '%s': invalid alias name\n", name);
+      }
+      else
+      {
+        /*
+          We now have a valid name and value, let add/update the
+          aliases hash.
+        */
+
+        if ((record= my_hash_search(&aliases, (const uchar *) name, name_len)))
+        {
+          my_hash_delete(&aliases, record);
+        }
+        alias= (ALIAS *) my_malloc(sizeof(ALIAS), MYF(MY_WME));
+        alias->name= name;
+        alias->name_len= name_len;
+        alias->value= value;
+        my_hash_insert(&aliases, (uchar *) alias);
+      }
+    }
+    else
+    {
+      /* Only alias name specified, lookup, print its value and return. */
+      alias= (ALIAS *) my_hash_search(&aliases, (const uchar*) name,
+                                      name_len);
+      if (alias)
+      {
+        tee_fprintf(stdout, "alias  %s = '%s'\n", alias->name, alias->value);
+      }
+      else
+      {
+        tee_fprintf(stdout, "alias: '%s': not found\n", name);
+      }
+    }
+    *error= false;
+  }
+  return pos;
+}
+
+static char *handle_next_unalias(char *line, bool *error)
+{
+  char *name, *pos;
+  uchar *record;
+  bool is_valid;
+
+  /* Parse the alias name. */
+  pos= parse_alias_name(line, &name, &is_valid);
+
+  if (!is_valid)
+  {
+    tee_fprintf(stdout, "unalias: '%s': invalid alias name\n", name);
+    *error= true;
+  }
+  else
+  {
+    if ((record= my_hash_search(&aliases, (const uchar *) name, strlen(name))))
+    {
+      my_hash_delete(&aliases, record);
+    }
+    else
+    {
+      tee_fprintf(stdout, "unalias: '%s': not found\n", name);
+    }
+    *error= false;
+  }
+  return pos;
+}
+static int init_alias()
+{
+  FILE *file= 0;
+  char *mariadbrc_file;
+  char *line= 0;
+  char *ptr;
+  size_t len= 0;
+  ssize_t read;
+  MY_STAT stat_arg;
+  bool error;
+  int pos;
+
+  /* Initialize the hash structure to store aliases. */
+  if (my_hash_init2(&aliases, 64, charset_info,
+                    64, 0, 0, get_alias_key, 0,
+                    alias_free, MYF(0)))
+  {
+    put_info("Couldn't initialize hash to store aliases.", INFO_ERROR);
+    exit(1);
+  }
+
+  /*
+    Read alias commands from mariadbrc file.
+  */
+  if (getenv("MARIADBRC_FILE"))
+  {
+    mariadbrc_file= my_strdup(getenv("MARIADBRC_FILE"), MYF(MY_WME));
```

Comment:
I believe it is worth to avoid calling ```getenv()``` twice and not calling ```strdup()``` at all. Something like:
```
char *mariadbrc_file;
char mariadbrc_buf[FN_REFLEN];

...

if (!(mariadbrc_file= getenv("MARIADBRC_FILE")))
{
  char *home= getenv("HOME");
  if (home)
  {
    snprintf(mariadbrc_buf, sizeof(mariadbrc_buf), "%s/.mariadbrc", home);
    mariadbrc_file= mariadbrc_buf;
  }
}
```